### PR TITLE
feat(first-event-severity): Add alert filter for issue severity

### LINKED ISSUE #54252: Create issue alert condition that filters on severity
We’ll make a condition here, called something like `HighSeverityIssueCondition`. The condition will filter to only new issues, and will check the severity of the group related to the event. Depending on how we decide to call the service, it’ll either fetch this from the group or directly call the service.

Initially, we’ll only want this condition to be available in the Sentry org. Here’s an example of how we’ve done this before https://github.com/getsentry/sentry/pull/26021/files

### PR DESCRIPTION
**Pull request by malwilley**
Closes https://github.com/getsentry/sentry/issues/54252

The ticket specified an alert condition, but it seemed to me like this was better suited to a filter instead. LMK if you disagree or I'm missing something there.

I added form elements so that these alerts are highly configurable and we can change the threshold as we wish. This will be useful for testing, but if we go ahead with this feature we probably want to hardcode thresholds and map them to low, med, high, etc.

This option will only be shown if the `projects:first-event-severity-alerting` feature flag is enabled.

CleanShot 2023-08-18 at 14 33 58

### CODE REVIEW

File: src/sentry/features/__init__.py [hunk 1]
```diff
@@ -101,6 +101,7 @@
 default_manager.add("organizations:issue-search-use-cdc-primary", OrganizationFeature, FeatureHandlerStrategy.REMOTE)
 default_manager.add("organizations:issue-search-use-cdc-secondary", OrganizationFeature, FeatureHandlerStrategy.REMOTE)
 default_manager.add("organizations:issue-search-group-attributes-side-query", OrganizationFeature, FeatureHandlerStrategy.REMOTE)
+default_manager.add("organizations:issue-severity-alerts", OrganizationFeature, FeatureHandlerStrategy.REMOTE)
```
Comment (scefali) [inline src/sentry/features/__init__.py:104]: @malwilley as long as this flag isn't used during post process then we can keep using flagr

Comment (malwilley) [reply src/sentry/features/__init__.py:104]: Yeah we'll use an option there

File: src/sentry/api/endpoints/project_rules_configuration.py [hunk 2]
```diff
@@ -64,7 +67,11 @@ def get(self, request: Request, project) -> Response:
             if rule_type.startswith("condition/"):
                 condition_list.append(context)
             elif rule_type.startswith("filter/"):
-                filter_list.append(context)
+                if (
+                    has_issue_severity_alerts
+                    or context["id"] != "sentry.rules.filters.issue_category.IssueSeverityFilter"
```
Comment (scefali) [inline src/sentry/api/endpoints/project_rules_configuration.py:72]: @malwilley I don't understand this logic, shouldn't it be:

`has_issue_severity_alerts and context["id"] == "sentry.rules.filters.issue_category.IssueSeverityFilter"`

Comment (malwilley) [reply src/sentry/api/endpoints/project_rules_configuration.py]: The logic does work here - added a test to confirm [`5b822b6` (#55095)](https://github.com/getsentry/sentry/pull/55095/commits/5b822b62e3b079960e35d954f8b23e447308d22a)

This is in a loop over all filters, so we want the condition to pass for everything except `not has_issue_severity_alerts and context["id"] = "...IssueSeverityFilter"`

**Review by scefali — APPROVED** [review]

**Comment by lobsterkatie** [comment]
FYI, hope you don't mind, I edited the PR description to talk about the new flag.

File: src/sentry/api/endpoints/project_rules_configuration.py [hunk 3]
```diff
@@ -64,7 +65,11 @@ def get(self, request: Request, project) -> Response:
             if rule_type.startswith("condition/"):
                 condition_list.append(context)
             elif rule_type.startswith("filter/"):
-                filter_list.append(context)
+                if (
+                    has_issue_severity_alerts
+                    or context["id"] != "sentry.rules.filters.issue_severity.IssueSeverityFilter"
+                ):
+                    filter_list.append(context)
```
Comment (lobsterkatie) [inline src/sentry/api/endpoints/project_rules_configuration.py:72]: Echoing Steve's earlier comment, the `context` condition might be a bit easier to grok in the affirmative:

```suggestion
                if (
                    context["id"] == "sentry.rules.filters.issue_severity.IssueSeverityFilter"
                    and not has_issue_severity_alerts
                ):
                    continue
                filter_list.append(context)
```

File: src/sentry/rules/filters/issue_severity.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    form_cls = IssueSeverityForm
+    form_fields = {
+        "value": {"type": "number", "placeholder": 0.5},
+        "match": {"type": "choice", "choices": list(SEVERITY_MATCH_CHOICES.items())},
+    }
+    rule_type = "filter/event"
+    label = "The issue's severity is {match} {value}"
+    prompt = "The issue's serverity is ..."
+
+    def _passes(self, group: Group):
```
Comment (lobsterkatie) [inline src/sentry/rules/filters/issue_severity.py:36]: Nit:
```suggestion
    def _passes(self, group: Group) -> bool:
```

File: src/sentry/rules/filters/issue_severity.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        has_issue_severity_alerts = features.has(
+            "projects:first-event-severity-alerting", self.project
+        )
+
+        if not has_issue_severity_alerts or not group:
```
Comment (lobsterkatie) [inline src/sentry/rules/filters/issue_severity.py:41]: Is it possible for `group` to be null? If so, it should probably be typed that way in the function signature.

File: src/sentry/rules/filters/issue_severity.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            return False
+
+        try:
+            severity = float(group.get_event_metadata().get("severity", ""))
+            value = float(self.get_option("value"))
```
Comment (lobsterkatie) [inline src/sentry/rules/filters/issue_severity.py:46]: Nit:
```suggestion
            threshold = float(self.get_option("value"))
```

For extra clarity. 🙂

File: src/sentry/rules/filters/issue_severity.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            return severity >= value
+        elif match == MatchType.LESS_OR_EQUAL:
+            return severity <= value
+
+        return False
+
+    def passes(self, event: GroupEvent, state: EventState) -> bool:
+        return self._passes(event.group)
+
+    def passes_activity(
```
Comment (lobsterkatie) [inline src/sentry/rules/filters/issue_severity.py:62]: Does this get covered by the existing tests?

File: tests/sentry/rules/filters/test_issue_severity.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+from sentry.rules.filters.issue_severity import IssueSeverityFilter
+from sentry.testutils.cases import RuleTestCase
+from sentry.testutils.helpers.features import with_feature
+
+
+class IssueSeverityFilterTest(RuleTestCase):
+    rule_cls = IssueSeverityFilter
+
+    @with_feature("projects:first-event-severity-alerting")
+    def test_valid_input_values(self):
```
Comment (lobsterkatie) [inline tests/sentry/rules/filters/test_issue_severity.py:12]: Nit:
```suggestion
    def test_passing_input_values(self):
```

"Invalid" makes me think we're testing against strings or `None` or something.

File: tests/sentry/rules/filters/test_issue_severity.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+        self.assertPasses(
+            self.get_rule(data={"match": MatchType.GREATER_OR_EQUAL, "value": 0.5}), event
+        )
+        self.assertPasses(
+            self.get_rule(data={"match": MatchType.GREATER_OR_EQUAL, "value": 0.7}), event
+        )
+        self.assertPasses(
+            self.get_rule(data={"match": MatchType.GREATER_OR_EQUAL, "value": "0.5"}), event
+        )
```
Comment (lobsterkatie) [inline tests/sentry/rules/filters/test_issue_severity.py:24]: This seems like a duplicate of the first one, no? Did you mean to have a second `LESS_OR_EQUAL` case?

File: tests/sentry/rules/filters/test_issue_severity.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        self.assertPasses(
+            self.get_rule(data={"match": MatchType.LESS_OR_EQUAL, "value": 0.9}), event
+        )
+
+    @with_feature("projects:first-event-severity-alerting")
+    def test_no_group_does_not_pass(self):
```
Comment (lobsterkatie) [inline tests/sentry/rules/filters/test_issue_severity.py:30]: Nit:
```suggestion
    def test_fail_on_no_group(self):
```

Just so it's consistent with the one below.

File: tests/sentry/rules/filters/test_issue_severity.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+            self.get_rule(data={"match": MatchType.GREATER_OR_EQUAL, "value": 0.5}), event
+        )
+
+    @with_feature("projects:first-event-severity-alerting")
+    def test_fail_on_no_severity(self):
+        event = self.get_event()
+
+        self.assertDoesNotPass(
+            self.get_rule(data={"match": MatchType.GREATER_OR_EQUAL, "value": 0.5}), event
+        )
```
Comment (lobsterkatie) [inline tests/sentry/rules/filters/test_issue_severity.py:45]: Might be worthwhile to also assert the lack of a severity score here.

File: tests/sentry/rules/filters/test_issue_severity.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+    @with_feature("projects:first-event-severity-alerting")
+    def test_fail_on_invalid_data(self):
```
Comment (lobsterkatie) [inline tests/sentry/rules/filters/test_issue_severity.py:48]: Nit:
```suggestion
    def test_failing_input_values(self):
```

To match the one above, and to handle the valid/invalid point raised above.

File: tests/sentry/rules/filters/test_issue_severity.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            {"match": MatchType.LESS_OR_EQUAL, "value": "0.5"},
+            {"match": MatchType.LESS_OR_EQUAL, "value": 0.5},
+            {"value": GroupCategory.ERROR.name},
+            {"match": MatchType.GREATER_OR_EQUAL},
+            {},
+        ]
+
+        for data_case in data_cases:
+            rule = self.get_rule(data=data_case)
+            self.assertDoesNotPass(rule, event)
```
Comment (lobsterkatie) [inline tests/sentry/rules/filters/test_issue_severity.py:64]: This is nice and neat. Perhaps do the test for passing values the same way?

**Review by lobsterkatie — COMMENTED** [review]
Looking good!

File: src/sentry/rules/filters/issue_severity.py (hunk 7 above)
Comment (malwilley) [reply src/sentry/rules/filters/issue_severity.py:62]: Good catch - added test coverage for this. Looks like this function isn't tested in any of the other rules, so I added a util function for it.

**Review by lobsterkatie — APPROVED** [review]
LGTM!

**Pull request merged**